### PR TITLE
feat: custom directives

### LINKED ISSUE #99: Adding new admonishes
Would be nice if new admonishes could be added by adding an icon and classes to the CSS.

I see that currently the directives are hardcoded.

I guess that would probably be a major change how this crate works.

### PR DESCRIPTION
**Pull request by tommilligan**
Closes #99
Replaces #157 
cc: @Sky9x as you've already thought a lot about this - I used a bunch of code and ideas from your branch, it was really helpful!

As I already wrote the doc page, I think this summarises the changes quite nicely:

image

---

The approach I took here was to write some slightly hacky Rust code to produce the small subset of CSS we need for each new variant. There's a unit test that will fail if either the Rust or SCSS code get out of step.

I think this is a good compromise between still doing the majority of complex CSS with SCSS, having runtime directive safety in Rust, and ensuring that there are some checks in place. An alternative would just have been to work out how to run an SCSS compiler inside a Rust binary, but I really did not want to do that.

This approach is also back compatible, and opt in for new users, which is great.

image

---

Internals wise, this changes slightly how the `Directive` type works internally. Basically, we standardise to a `String` earlier in the pipeline, at the point we resolve the `title`.

Uses `hex_color` for that concern as it's simple and looks to do the job well.

### CODE REVIEW

File: integration/expected/book.toml [hunk 1]
```diff
@@ -9,13 +9,18 @@ title = "mdbook-admonish-integration"
...
 command = "mdbook-admonish"
-assets_version = "3.0.1" # do not edit: managed by `mdbook-admonish install`
+assets_version = "3.0.2" # do not edit: managed by `mdbook-admonish install`
 after = ["links"]
 
+[[preprocessor.admonish.custom]]
+directive = "frog"
+icon = "./frog.svg"
+color = "#9004CC"
+
```
Comment (Sky9x) [inline integration/expected/book.toml:19]: <3

**Comment by Sky9x** [comment]
this looks cool! and better than what i threw together at 5am :sleepy:

File: integration/expected/book.toml (hunk 1 above)
Comment (tommilligan) [reply integration/expected/book.toml:19]: love a good :rainbow_flag: :frog:

File: book/src/reference.md [hunk 2]
```diff
@@ -75,6 +75,31 @@ Subfields:
...
+aliases = ["money", "cash", "budget"]
+```
+
+Subfields:
+
+- `directive`: The keyword to use this type of block.
+- `icon`: A filepath relative to the book root to load an SVG icon from.
+- `color`: An RGB hex encoded color to use for the icon.
+- `aliases` (optional): One or more alternative directives to use this block.
+- `title` (optional): The default title for this type of block.
```
Comment (Sky9x) [inline book/src/reference.md]:
```suggestion
- `title` (optional): The default title for this type of block. If not specified, defaults to the directive in title case.
```

File: book/src/reference.md [hunk 3]
```diff
@@ -75,6 +75,31 @@ Subfields:
...
+color = "#24ab38"
+aliases = ["money", "cash", "budget"]
+```
+
+Subfields:
+
+- `directive`: The keyword to use this type of block.
+- `icon`: A filepath relative to the book root to load an SVG icon from.
+- `color`: An RGB hex encoded color to use for the icon.
+- `aliases` (optional): One or more alternative directives to use this block.
```
Comment (Sky9x) [inline book/src/reference.md:100]: how do aliases interact with custom default titles?

File: Cargo.toml [hunk 4]
```diff
@@ -45,6 +45,7 @@ serde_json = "1.0.107"
 toml_mdbook = { package = "toml", version = "0.5.11" }
 toml = "0.8.1"
 toml_edit = { version = "0.20.1", optional = true }
+hex_color = { version = "3.0.0", features = ["serde"] }
```
Comment (Sky9x) [inline Cargo.toml:49]: a library! much better than my hack

**Review by Sky9x — APPROVED** [review]

File: book/src/reference.md (hunk 3 above)
Comment (Sky9x) [reply book/src/reference.md:100]: i had worked around this problem originally by having aliases just not exist and instead duplicating the custom directives for each alias

Comment (tommilligan) [reply book/src/reference.md:100]: I'll document - the same title would apply to all aliases. Basically, if you want different titles, just add two custom blocks

Comment (Sky9x) [reply book/src/reference.md:100]: :+1:

**Pull request merged**